src/ray/raylet/scheduling/policy/node_label_scheduling_policy.cc

```diff
@@ -0,0 +1,195 @@
+// Copyright 2023 The Ray Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "ray/raylet/scheduling/policy/node_label_scheduling_policy.h"
+
+#include "ray/raylet/scheduling/policy/scheduling_context.h"
+
+namespace ray {
+namespace raylet_scheduling_policy {
+
+scheduling::NodeID NodeLabelSchedulingPolicy::Schedule(
+    const ResourceRequest &resource_request, SchedulingOptions options) {
+  RAY_CHECK(options.scheduling_type == SchedulingType::NODE_LABEL);
+  auto context =
+      dynamic_cast<const NodeLabelSchedulingContext *>(options.scheduling_context.get());
+  const auto &scheduling_strategy = context->GetSchedulingStrategy();
+  RAY_CHECK(scheduling_strategy.has_node_label_scheduling_strategy());
+  const auto &node_label_scheduling_strategy =
+      scheduling_strategy.node_label_scheduling_strategy();
+  // 1. Selet feasible nodes
+  auto candidate_nodes = SelectFeasibleNodes(resource_request);
+  if (candidate_nodes.empty()) {
+    return scheduling::NodeID::Nil();
+  }
+  // 2. Filter by hard expressions.
+  if (node_label_scheduling_strategy.hard().expressions().size() > 0) {
+    SelectNodesByHardExpressions(candidate_nodes, node_label_scheduling_strategy.hard());
+    if (candidate_nodes.empty()) {
+      return scheduling::NodeID::Nil();
+    }
+  }
+
+  // 3. Filter by soft expressions.
+  const auto &soft_expressions = node_label_scheduling_strategy.soft();
+  if (soft_expressions.expressions().size() > 0) {
+    auto match_all_soft_nodes =
+        SelectNodesBySoftExpressions(candidate_nodes, soft_expressions);
+    if (!match_all_soft_nodes.empty()) {
+      return SelectBestNode(match_all_soft_nodes, resource_request);
+    }
+  }
+
+  return SelectBestNode(candidate_nodes, resource_request);
+}
+
+scheduling::NodeID NodeLabelSchedulingPolicy::SelectBestNode(
+    const absl::flat_hash_map<scheduling::NodeID, const Node *> &candidate_nodes,
+    const ResourceRequest &resource_request) {
+  auto available_nodes = SelectAvailableNodes(candidate_nodes, resource_request);
+  if (available_nodes.empty()) {
+    return SelectRandomNode(candidate_nodes);
```

Comment:
I think we can start with random but I can see in the future we will use hybrid policy or spread policy to select the best node among candidates.